interfaces/ITextToSpeech.h

```diff
@@ -0,0 +1,121 @@
+#ifndef __ITEXTTOSPEECH_H
+#define __ITEXTTOSPEECH_H
+
+#include "Module.h"
+
+namespace WPEFramework {
+namespace Exchange {
+
+    struct EXTERNAL ITextToSpeech : virtual public Core::IUnknown {
+        enum { ID = ID_TEXT_TO_SPEECH };
+
+        struct Configuration {
+            std::string ttsEndPoint;
+            std::string ttsEndPointSecured;
+            std::string language;
+            std::string voice;
+            uint8_t volume;
+            uint8_t rate;
+        };
+
+        enum SpeechState: uint8_t {
+            SPEECH_PENDING = 0,
+            SPEECH_IN_PROGRESS,
+            SPEECH_PAUSED,
+            SPEECH_NOT_FOUND
+        };
+
+        enum TTSErrorDetail: uint8_t {
+            TTS_OK = 0,
+            TTS_FAIL,
+            TTS_NOT_ENABLED,
+            TTS_INVALID_CONFIGURATION,
+            TTS_NO_ACCESS
+        };
+
+
+        struct INotification : virtual public Core::IUnknown {
+            enum { ID = ID_TEXT_TO_SPEECH_NOTIFICATION };
+
+            ~INotification() override = default;
+
+            // Signal changes on the subscribed namespace
+            // @brief Notify TTS enabled/disabled
+            // @param state enabled/disabled
+            virtual void Enabled(const bool state) = 0;
+
+            // @brief Notify change in voice used for speaking
+            // @param voice voice changed 
+            virtual void VoiceChanged(const string voice) = 0;
+
+            // @brief Notify speechid based on the speech state(eg: start,pause,..etc)
+            // @param speechid id of the text
+            virtual void WillSpeak(const uint32_t speechid) = 0;
+            virtual void SpeechStart(const uint32_t speechid) = 0;
+            virtual void SpeechPause(const uint32_t speechid) = 0;
+            virtual void SpeechResume(const uint32_t speechid) = 0;
+            virtual void SpeechInterrupted(const uint32_t speechid) = 0;
+            virtual void NetworkError(const uint32_t speechid) = 0;
+            virtual void PlaybackError(const uint32_t speechid) = 0;
+            virtual void SpeechComplete(const uint32_t speechid) = 0;
+        };
+
+        ~ITextToSpeech() override = default;
+
+        virtual void Register(ITextToSpeech::INotification* sink) = 0;
+        virtual void Unregister(ITextToSpeech::INotification* sink) = 0;
+        
+        // @property
+        // @brief Query the status/enable tts
+        // @param enable status/enable
+        virtual uint32_t Enable(const bool enable) = 0;
+        virtual uint32_t Enable(bool &enable /* @out */) const = 0;
+
+        // @brief Set the tts configuration attributes
+        // @param config tts configuration
+        // @param status return status
+        virtual uint32_t SetConfiguration(const Configuration config,TTSErrorDetail &status/* @out */) = 0;
```

Comment:
The Configuration type is an object (struct), passing it as a reference, avoids a copy (depending on compiler optimizations). As stated in the previous review, only scalars that are const should not be passed by reference. Objects (structs) should always be passed by reference to avoid expensive copies.

Reply:
updated